### PR TITLE
feat(security): close DNS-rebinding gap in HTTP client (CURLOPT_RESOLVE)

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary

Closes the documented DNS-rebinding gap in our SSRF defence. Previously, `isHostAllowed()` resolved + validated the host at check time, but curl re-resolved the same hostname at connect time — an attacker with DNS control could return a benign IP on the first lookup (passing our check) and an internal IP on the second (where curl actually connects). TOCTOU via the resolver.

This is the follow-up flagged in the `isDangerousIpLiteral()` docblock since the H-1 SSRF defence landed:

> *"Caveat: this defence is bypassable by DNS rebinding when the upstream HTTP client (Guzzle/curl) re-resolves at connect-time. For full protection, callers must pin to the resolved IP via curl `CURLOPT_RESOLVE`; that is a follow-up."*

## The fix

Push a Guzzle middleware (`ssrf-dns-pin`) onto the HandlerStack inside `SecureHttpClientFactory::create()`. The middleware runs per outgoing request and:

1. Reads URI host + port.
2. Resolves via `dns_get_record(A | AAAA)` and validates EACH record against the existing `isDangerousIpLiteral()` defence.
3. **Any** dangerous answer kills the request → `RequestException` before socket opens. (Split-horizon rebinding could otherwise return one safe + one internal IP; if curl picked the internal one we'd leak.)
4. All-safe records get pinned via curl's `CURLOPT_RESOLVE` (`host:port:ip` entries, supports multiple for dual-stack). curl then skips its own DNS step.
5. IP literals (already validated) and unresolvable hosts (let curl produce the error) pass through without a pin.

## Architecture

| Layer | Purpose | When |
|---|---|---|
| `VaultHttpClient::sendRequest()` → `isHostAllowed()` | Early gate; happens BEFORE secret injection so credentials never touch a rejected request | Per request, our code |
| `ssrf-dns-pin` middleware | Late gate; pins the verified IP so curl can't be tricked by a second resolution | Per request, in Guzzle's request pipeline |

Both checks run; the middleware adds the race-free pin on top of the existing allowlist defence.

## Why we can do this

We instantiate `GuzzleHttp\Client` **directly** (see PR description of the discussion: we use our own HTTP client, not TYPO3's `RequestFactory`). That gives us full access to curl options. A factory built on TYPO3's `GuzzleClientFactory` would not have this option.

## Test plan

- [x] +5 unit tests in `Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php`:
  - `buildResolveEntriesReturnsEmptyForIpLiteral` — IP literals don't get a pin.
  - `buildResolveEntriesReturnsEmptyForUnresolvableHost` — `.invalid` TLD → curl handles connect-failure.
  - `middlewareIsRegisteredInTheHandlerStack` — the `ssrf-dns-pin` name appears on the stack.
  - `middlewareDoesNotAddCurlResolveForIpLiteralHost` — full round-trip with captured bottom handler.
  - `middlewareRejectsRequestToHostResolvingToLoopback` — `localhost` (→ 127.0.0.1) triggers the dangerous-IP path; `RequestException` with the rebinding-defence message.
- [x] 1717 unit tests pass (+5 new).
- [x] cgl + rector + PHPStan level 10 green locally.
- [ ] CI matrix.

## Not in scope

- `OAuthTokenManager` has its OWN `GuzzleHttp\Client` instance bypassing this factory. That's a separate gap and deserves its own PR (the manager should accept an injected client from `SecureHttpClientFactory` instead of constructing one). Will track.
- IPv6 dual-stack edge cases: middleware pins both A and AAAA records when present, but TLS SNI + curl's per-family interface selection may still surprise. Functional tests against real IPv6-only hosts would help; deferred.

## Migration

No behaviour change for legitimately-public targets. Rejects new (correct) cases:
- Hostnames whose DNS records include private/loopback/metadata IPs.
- Rebinding-style domain setups that try to swap resolution between check and connect.

If a customer has an on-prem deployment where the vault server lives on RFC1918, they need the existing `$GLOBALS['TYPO3_CONF_VARS']['HTTP']['allowed_hosts']` allowlist (literal match) — that path is unchanged and still bypasses the IP guard.

### CODE REVIEW

File: Classes/Http/SecureHttpClientFactory.php [hunk 1]
```diff
@@ -174,6 +183,120 @@ public function isHostAllowed(string $host): bool
...
+     *    earlier `isHostAllowed()` check already rejected IP-literal targets
+     *    in dangerous ranges.)
+     *  - IP-literal hosts (already an IPv4/IPv6 address) → no pin needed;
+     *    `isHostAllowed()` validated the literal directly.
+     */
+    private function buildSsrfDefenceMiddleware(): Closure /** @phpstan-ignore missingType.callable (Guzzle's middleware contract is loosely-typed by design) */
+    {
+        return function (callable $handler): Closure { /** @phpstan-ignore missingType.callable */
+            return function (RequestInterface $request, array $options) use ($handler) {
+                $host = $request->getUri()->getHost();
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Http/SecureHttpClientFactory.php]: Excellent catch — real SECURITY bug. Fixed in '520199b': the middleware now passes the host through `normaliseHost()` before any IP check or DNS lookup, so bracketed IPv6 forms (`[::1]`) are handled correctly. Added `middlewareNormalisesBracketedIpv6Host` as a regression test.

File: Classes/Http/SecureHttpClientFactory.php [hunk 2]
```diff
@@ -174,6 +183,120 @@ public function isHostAllowed(string $host): bool
...
+            if (!\is_string($ip)) {
+                continue;
+            }
+            if ($this->isDangerousIpLiteral($ip)) {
+                // ANY dangerous answer kills the entire request. A "split-horizon"
+                // rebinding setup could otherwise return one safe + one internal
+                // IP; if curl picked the internal one, we'd leak.
+                return null;
+            }
+            $entries[] = \sprintf('%s:%d:%s', $host, $port, $ip);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Http/SecureHttpClientFactory.php]: Fixed in '520199b' — IPv6 addresses are now bracketed in CURLOPT_RESOLVE entries (`str_contains($ip, ':') ? '[' . $ip . ']' : $ip`). Added `buildResolveEntriesBracketsIpv6` as a regression test.

File: Classes/Http/SecureHttpClientFactory.php [hunk 3]
```diff
@@ -100,6 +103,12 @@ public function create(): ClientInterface
 
         // Create handler stack without any logging middleware
         $stack = HandlerStack::create();
+
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Http/SecureHttpClientFactory.php:110]: Fixed in '520199b' — `create()` now logs a warning when ext-curl is missing so operators notice the gap. Chose log-and-degrade over hard-fail because the middleware's pre-request validation (resolve + check) still fires on StreamHandler — it's just no longer race-protected against rebinding. Hard-fail would break legitimate deployments that don't ship ext-curl; the warning makes the security trade-off visible.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+/*
+ * Copyright (c) 2025-2026 Netresearch DTT GmbH
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+declare(strict_types=1);
+
+namespace Netresearch\NrVault\Tests\Unit\Http;
+
+use GuzzleHttp\Promise\PromiseInterface;
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php]: Fixed in '520199b' — introduced `DnsResolverInterface` + `DefaultDnsResolver` (production) + `InMemoryDnsResolver` test double. The Factory constructor injects the resolver (defaults to DefaultDnsResolver so existing makeInstance call sites still work). Tests now program responses per host and assert on the recorded queries — no real DNS. Bonus: the existing `resolvesToDangerousIp` site also delegates through the resolver, so the entire SSRF defence is testable without flakiness.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            unset($GLOBALS['TYPO3_CONF_VARS']);
+        }
+    }
+
+    #[Test]
+    public function buildResolveEntriesReturnsEmptyForIpLiteral(): void
+    {
+        // IPv4 and IPv6 literals are validated by isHostAllowed() — the
+        // middleware doesn't need to add a pin entry.
+        self::assertSame([], $this->callBuildResolveEntries('93.184.216.34', 443));
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:73]: Test-only finding — these are intentionally hardcoded IPs used as DNS-rebinding test programming (one safe public, one AWS-metadata loopback). Same category as the bulk-accepted test-noise findings in PR #140's UI cleanup. Will mark accepted with rationale once Sonar re-analyzes.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        );
+    }
+
+    #[Test]
+    public function middlewareDoesNotAddCurlResolveForIpLiteralHost(): void
+    {
+        $client = $this->buildCapturingClient($capturedOptions);
+
+        // IP literal — buildResolveEntries returns [] and no pin is added.
+        $client->get('http://93.184.216.34/');
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:145]: Test-only finding — these are intentionally hardcoded IPs used as DNS-rebinding test programming (one safe public, one AWS-metadata loopback). Same category as the bulk-accepted test-noise findings in PR #140's UI cleanup. Will mark accepted with rationale once Sonar re-analyzes.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        // usual connection-failure error path.
+        $entries = $this->callBuildResolveEntries('unknown.example', 443);
+
+        self::assertSame([], $entries);
+    }
+
+    #[Test]
+    public function buildResolveEntriesPinsSafeIpv4(): void
+    {
+        $this->dnsResolver->program('api.example.com', [['ip' => '93.184.216.34']]);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:90]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    #[Test]
+    public function buildResolveEntriesReturnsNullWhenAnyRecordIsDangerous(): void
+    {
+        // Split-horizon: resolver returns one public + one internal IP. ANY
+        // dangerous answer must kill the request — curl could otherwise pick
+        // the internal one and we'd leak.
+        $this->dnsResolver->program('rebind.example.com', [
+            ['ip' => '93.184.216.34'],
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:116]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            ['ip' => '169.254.169.254'], // AWS metadata
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:117]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        // IP literal — buildResolveEntries returns [] and no pin is added.
+        $client->get('http://93.184.216.34/');
+
+        self::assertArrayNotHasKey('curl', $capturedOptions ?? []);
+    }
+
+    #[Test]
+    public function middlewareRejectsRequestToHostResolvingToDangerousIp(): void
+    {
+        $this->dnsResolver->program('attacker.example', [['ip' => '169.254.169.254']]);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:153]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        $this->expectException(RequestException::class);
+        $this->expectExceptionMessageMatches('/DNS rebinding defence/i');
+
+        $client->get('http://attacker.example/');
+    }
+
+    #[Test]
+    public function middlewareAddsCurlResolvePinForResolvedHost(): void
+    {
+        $this->dnsResolver->program('safe.example', [['ip' => '93.184.216.34']]);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:165]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.

File: Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        // PSR-7 getHost() returns IPv6 literals wrapped in brackets. Without
+        // normalisation, `filter_var(..., FILTER_VALIDATE_IP)` would reject
+        // `[::1]` and the middleware would try to resolve it via DNS — both
+        // useless and a security regression (the IP-literal guard wouldn't
+        // run). normaliseHost() strips the brackets first.
+        //
+        // Verified indirectly: send to a bracketed FQDN-style host that
+        // resolves via the in-memory resolver. If normalisation runs, the
+        // resolver receives the bare host and the pin is generated.
+        $this->dnsResolver->program('safe.example', [['ip' => '93.184.216.34']]);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Http/SecureHttpClientFactoryRebindingTest.php:188]: Test fixture — synthetic IPs used to program the InMemoryDnsResolver and exercise the SSRF / DNS-rebinding defence: 93.184.216.34 (synthetic-public), 169.254.169.254 (AWS metadata loopback used in the rebinding-rejection test), 2001:db8::1 (RFC 3849 documentation prefix). All inside Tests/Unit/; production code never references these literals.
